src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs

```diff
@@ -144,6 +149,14 @@ public void ServiceProviderBuilt(ServiceProvider provider)
         {
             lock (_providers)
             {
+                int providersCount = _providers.Count;
+                if (providersCount > 0 &&
+                    (_survivingProvidersCount is int spc ? (uint)providersCount > 2 * (uint)spc : providersCount == _providers.Capacity))
+                {
+                    _providers.RemoveAll(static p => !p.TryGetTarget(out _));
+                    _survivingProvidersCount = _providers.Count;
+                }
+
```

Comment:
Now that the list is being culled when adding, in ServiceProviderDisposed should we break out of the loop once we find target == provider?

Reply:
If we are preoccupied about duplication, we're already cleaning in two places. While at it, I prefer doing it aggressively and not halfway

Reply:
Why? What is the benefit?

Reply:
At the expense of an O(n) iteration on every dispose, which means creating/disposing n providers can be o(n^2). All iterating the full list does then on Dispose is allow some empty weak refs to be cleaned up. If every provider is disposed of properly, looking for empty weak refs will never be fruitful. If disposed is never used, this code doesn't matter. So we're only taking about situations where lots are created and only some are disposed.